### ISSUE TITLE
core.setOutput should(?) transform `data` to a string

### ISSUE THREAD

**Issue by JasonEtco** [actions/toolkit#370]
**Describe the enhancement**

Hello! In an action I maintain, I call `setOutput` with an issue number. That fails, because `setOutput` expects it to be a string. This bit me (https://github.com/JasonEtco/create-an-issue/issues/46), and I'm wondering if y'all would be up for transforming the value passed to a string, or throwing if it isn't a string.

I realize that this gets into "just use TypeScript" territory, but I see no reason that I shouldn't be able to pass a number aside from the implementation detail of how it works under the hood.

**Code Snippet**

```js
// Doesn't work but could
core.setOutput('issue_number', 1)

// Does work
core.setOutput('issue_number', '1')
core.setOutput('issue_number', String(1))
```

**Additional information**

I'm using [JasonEtco/actions-toolkit](https://github.com/JasonEtco/actions-toolkit) as an opinionated alternative/wrapper for `actions/toolkit`. If transforming the value isn't something y'all feel good about, I'm happy to move that into that more opinionated territory!

**Comment by ericsciple** [comment]
@thboop thoughts?

I think this makes sense. Especially given common cases like bool/number.

The only thing that worries me a little bit, is if we ever preserve type/structure in the future. Basically to enable something like `${{ steps.foo.outputs.bar.subproperty }}`. However we could always overload in the future... Also it's nice to just keep it simple (outputs are strings), so i think converting to string makes sense.

I wonder for arrays/objects, whether we should convert to json (indented) or just defer to however nodejs converts to string - for example:
```js
`${[1, 2, 3]}` // '1,2,3'
```

i kind of like just doing whatever nodejs does (no magic)

**Comment by JasonEtco** [comment]
Thanks for sharing those thoughts @ericsciple, there are a couple of complications you mentioned.

> I wonder for arrays/objects, whether we should convert to json (indented?)

This is an interesting choice; using the `String` primitive on an object, it does this:

```js
String({ test: true })
// -> "[object Object]"
```

Which is not at all useful. This is way better:

```js
JSON.stringify({ test: true })
// -> '{ "test": true }'
```

`JSON.stringify` can also be applied numbers, arrays an booleans - so that might be a nice catch-all approach.

**Comment by thboop** [comment]
I think stringifying makes a lot of sense. I'd love to make this more usable for bools and numbers.

`Also it's nice to just keep it simple (outputs are strings), so i think converting to string makes sense.` Agreed.

**Comment by JasonEtco** [comment]
One thing to note is that there may be an expectation that whatever type you `setOutput` will be the same type on the other end - that sounds like something for docs (and obviously TypeScript).

**Comment by ericsciple** [comment]
+1 i would just write in the function description that it runs `JSON.stringify` for any values that arent a string

some quirks to be aware of :)
```js
JSON.stringify(null) // 'null'
JSON.stringify(undefined) // undefined - so need to explicitly handle undefined
```

**Comment by Jetersen** [comment]
Feel into this trap: https://github.com/release-drafter/release-drafter/pull/376#discussion_r392698467

**Comment by thboop** [comment]
Resolved in #405

**Comment by nandorojo** [comment]
Question, what if I want to `setOutput` as an array?

I've been trying to do this:

```js
const output = JSON.stringify(['live', 'sandbox'])
core.setOutput('environments', output)
```

And then I want to read this in my action with `fromJSON()`:

```yml
  matrix:
    app-env: ${{fromJSON(needs.environments.outputs.environments)}}
```

However, it seems that `fromJSON` doesn't like the stringified JSON provided by `JSON.stringify`. Is there a way to parse this properly?
